url0e7d57473a_kyorohiro_HetimaUtil_tgz-pJ8-net_hetimatan_net_http_HttpServer3xxJ8/src/net/hetimatan/util/event/EventTaskRunnerImple.java

```diff
@@ -0,0 +1,215 @@
+		return mRunner.isAlive();
+	}
+
+	public synchronized void pushTask(EventTask task) {
+		start(task);
+	}
+
+	public synchronized void pushTask(EventTask task, int timeout) {
+		mDefferTasks.add(new DefferTask(task, timeout));
+		if(mRunner == null|| mWorker == null) {
+			start(null);
+		}
+		kickWorker();
+	}
+
+	@Override
+	public void releaseTask(EventTask task) {
+		if(task != null) {
+		mDefferTasks.remove(task);
+		}
+	}
+
+	public int updateDeffer() {
+		long curTime = System.currentTimeMillis();
+		long time = -1;
+		long ret = -1;
+		for(int i=0;i<mDefferTasks.size();i++) {
+			DefferTask task = mDefferTasks.get(i);
+			time = task.deffer(curTime);
+			if(time<=0) {
+				pushTask(task.getEventTask());
+				mDefferTasks.remove(i);
+				ret = 0;
+			} else {
+				if(ret==-1) {
+					ret = time;					
+				}
+				else if(ret>time) {
+					ret = time;
+				}
+			}
+		}
+		return (int)ret;
+	}
+
+	public synchronized EventTask popTask() {
+		try {
+		if (mTasks.size() > 0) {
+			return mTasks.remove(0);//removeFirst();
+//			return mTasks.removeFirst();//(0);//removeFirst();
+		} else {
+			return null;
+		}
+		} catch(NoSuchElementException e) {
+			return null;
+		}
+	}
+
+	@Override
+	public void start(EventTask task) {
+		super.start(task);
+		//pushWork(task);
+//		mTasks.addLast(task);
+		if (task != null) {
+			mTasks.add(task);
+		}
+		if (mRunner == null || !mRunner.isAlive()) {
+			mRunner = new SingleTaskRunner();
+			mRunner.startTask(mWorker = new Worker(this));
+		} else if(mWorker != null){
+			//mWorker.kick();
+			kickWorker();
+		}
+	}
+
+	public void dispose() {
+		mRunner.endTask();
+	}
+
+	public void kickWorker() {
+		mWorker.kick();
+	}
+
+	public synchronized void kick() {
+		notify();
+	}
+	public synchronized void waitPlus(int time) throws InterruptedException, IOException {
+		if(time<0) {
+			//Thread.sleep(10000);
+			wait(10000);
+		} else {
+			if(time > 0) {
+				wait(time);
+			}
+			//Thread.sleep(time);			
+		}
+	}
+
+	public static class Worker implements Runnable {
+
+		WeakReference<EventTaskRunnerImple> mRunner = null;
+
+		public Worker(EventTaskRunnerImple runner) {
+			mRunner = new WeakReference<EventTaskRunnerImple>(runner);
+		}
+
+		public void kick() {
+			EventTaskRunnerImple runner = mRunner.get();
+			runner.kick();
+		}
+
+		@Override
+		public void run() {
+			putWorker(Thread.currentThread(), mRunner.get());
+			try {
+				while (true) {
+					if(mRunner == null) {break;}
+					EventTaskRunnerImple runner = mRunner.get();
+					if (null == runner||runner.mRunner == null) {
+						break;
+					}
+					EventTask task = runner.popTask();
+					if (task == null) {
+						//synchronized(runner){
+							synchronized(this){
+								if(!Thread.interrupted()) {
+									runner.waitPlus(runner.updateDeffer());
+								} 
+							}
+						//}
+					} else {
+						task.run(mRunner.get());
+					}
+				}
+			} catch(InterruptedException e) {} catch(IOException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)